### ISSUE TITLE
OAuth failure with Twitter

### ISSUE THREAD

**Issue by mbrung** [chok/sfMelodyPlugin#8]
Hi !

I seem to be having difficulties in authenticating with Twitter.  While the FB authentication works like a charm, Twitter auth fails upon callback.

After the user allows my app to access his/her info, Twitter calls us back with the following URL:

access/twitter?oauth_token=7WejrvRcKrUQxF16QAhb525O128146CTTA1K5tW4&oauth_verifier=hzOEQWtg34tPr4YG8NMRm1Y6GNDKYBishrK88aE

Seems legit so far.

Calling this URL fails, and I see this message in my log:

Sep 22 12:33:33 symfony [err] {OAuth} access token failed -  returns Array
(
    [ ] => 
)

I'm a bit at a loss here -- I'm sure someone else got this to work, so I wonder if perhaps I have some configuration problem.

Would you have any suggestions/ideas?

FYI the _twitter_ section in my app.yml looks like:

```
twitter:
  provider: twitter
  key: <REMOVED>
  secret: <REMOVED>
  callback: @homepage_twitter
  scope: [email]

  user:
    username:               
      call: me         
      path: id                
      prefix: "Twitter_"
    first_name:
      call: me
      path: first_name
    last_name: 
      call: me
      path: last_name
    email_address:               
      call: me                  
      path: email               
      key: true
```

**Comment by chok** [comment]
There is no scope with twitter. And You can't retrieve the email of the user.

me is an alias to call :this : http://dev.twitter.com/doc/get/users/show

As you can see in the example, there is no first_name, last_name or email_address.

**Comment by mbrung** [comment]
Thanks.

So I stripped down the twitter section in my app.yml:

```
twitter:
  provider: twitter
  key: <REMOVED>
  secret: <REMOVED>
  callback: @homepage_twitter

  user:
    username:            
      call: me          
      path: id     
      prefix: "Twitter_"
      key: true
```

Unfortunately I get the same result:

  Sep 22 13:16:15 symfony [info] {sfMelodyActions} Call "sfMelodyActions>executeAccess()"
  Sep 22 13:16:16 symfony [err] {OAuth} access token failed -  returns Array
  (
      [ ] => 
  )

**Comment by chok** [comment]
Can you post here all lines about OAuth in your log file ?

And there is a lin like this : Sep 22 21:01:46 symfony [info] {sfPatternRouting} Match route "melody_access" (/access/:service) for /access/twitter with parameters array (  'module' => 'sfMelody',  'action' => 'access',  'service' => 'twitter',) ?

In fact, another person have the probleme without find solution.

So, thanks to help to solve this problem :)

**Comment by mistrymj** [comment]
is there a solution to this?

I am having a similar problem... after I try to do a connect to twitter, the browser is at https://api.twitter.com/oauth/authorize?

which is displays "Woah there!

This page requires some information that was not provided. Please return to the site that sent you to this page and try again … it was probably an honest mistake.
"

from my frontend_dev.log I have...
Mar 27 15:21:37 symfony [info] User agent 'Mozilla/5.0 (Macintosh; U; Intel Mac OS X 10.6; en-US; rv:1.9.2.16) Gecko/20110319 Firefox/3.6.16
        ' detected. Selected device: html
Mar 27 15:21:37 symfony [info] {sfPatternRouting} Connect sfRoute "sf_guard_signin" (/guard/login)
Mar 27 15:21:37 symfony [info] {sfPatternRouting} Connect sfRoute "sf_guard_signout" (/guard/logout)
Mar 27 15:21:37 symfony [info] {sfPatternRouting} Match route "default" (/:module/:action/*) for /connect/twitter with parameters array (  'module' => 'connect',  'action' => 'twitter',)
Mar 27 15:21:37 symfony [info] {sfFilterChain} Executing filter "sfRenderingFilter"
Mar 27 15:21:37 symfony [info] {sfFilterChain} Executing filter "sfGuardRememberMeFilter"
Mar 27 15:21:37 symfony [info] {sfFilterChain} Executing filter "sfExecutionFilter"
Mar 27 15:21:37 symfony [info] {connectActions} Call "connectActions->executeTwitter()"
Mar 27 15:21:37 symfony [info] {OAuth} call url "https://api.twitter.com/oauth/request_token" with params "oauth_callback=http%3A%2F%2Flocal.roundonme.com%2Ffrontend_dev.php%2Faccess%2Ftwitter&oauth_consumer_key={my KEY}&oauth_nonce=becffa2a232d5af19c81deaa200fa875&oauth_signature=%2FMfbnpbOtdqO%2BMhPH34A%2BF8mZMQ%3D&oauth_signature_method=HMAC-SHA1&oauth_timestamp=1301253697&oauth_version=1.0" post "" by method "POST" give response ""
Mar 27 15:21:37 symfony [err] {OAuth} access token failed - twitter returns Array
(
)

Mar 27 15:21:37 symfony [info] {OAuth} "twitter" call url "https://api.twitter.com/oauth/authorize" with params "array (
  'oauth_token' => NULL,
)"
Mar 27 15:21:37 symfony [info] {sfFrontWebController} Redirect to "https://api.twitter.com/oauth/authorize?"
Mar 27 15:21:37 symfony [info] {sfWebResponse} Send status "HTTP/1.1 302 Found"
Mar 27 15:21:37 symfony [info] {sfWebResponse} Send header "Location: https://api.twitter.com/oauth/authorize?"
Mar 27 15:21:37 symfony [info] {sfWebResponse} Send header "Content-Type: text/html; charset=utf-8"
Mar 27 15:21:37 symfony [info] {sfWebDebugLogger} Configuration 55.32 ms (8)
Mar 27 15:21:37 symfony [info] {sfWebDebugLogger} Factories 81.40 ms (1)
Mar 27 15:21:37 symfony [info] {sfWebDebugLogger} Action "connect/twitter" 217.67 ms (1)
Mar 27 15:21:37 symfony [info] {sfWebResponse} Send content (111 o)
Mar 27 15:21:37 symfony [err] {sfRenderException} The template "twitterSuccess.php" does not exist or is unreadable in "".
Mar 27 15:21:38 symfony [info] {sfWebDebugLogger} Configuration 158.67 ms (9)
Mar 27 15:21:38 symfony [info] {sfWebDebugLogger} Factories 81.40 ms (1)
Mar 27 15:21:38 symfony [info] {sfWebDebugLogger} Action "connect/twitter" 435.95 ms (2)
Mar 27 15:21:38 symfony [info] {sfWebDebugLogger} View "Success" for "connect/twitter" 254.74 ms (1)

Any ideas?

I also have a similar problem trying to get the Facebook connect to work...

Mar 27 15:26:07 symfony [info] {OAuth} call url "https://graph.facebook.com/oauth/access_token?client_id=216991862191&client_secret=342cf49a1d9b6e682b93efe40b499355&redirect_uri=http%3A%2F%2Flocal.site.com%2Ffrontend_dev.php%2Faccess%2Ffacebook&code=2.84A0u65SC1mQYLcom3NhrQ__.3600.1301389200-689370393%7CqpVZEXAw3y5jjxP04RVzvOQ364Q" with params "client_id={KEY}&client_secret={MY SECRET}&redirect_uri=http%3A%2F%2Flocal.site.com%2Ffrontend_dev.php%2Faccess%2Ffacebook&code=2.84A0u65SC1mQYLcom3NhrQ__.3600.1301389200-689370393%7CqpVZEXAw3y5jjxP04RVzvOQ364Q" post "" by method "GET" give response ""
Mar 27 15:26:07 symfony [err] {OAuth} access token failed - facebook returns Array
(
)

Mar 27 15:26:08 symfony [info] {OAuth} call url "https://graph.facebook.com/me" with params "" post "" by method "GET" give response ""
Mar 27 15:26:08 symfony [info] {OAuth} call url "https://graph.facebook.com/me" with params "" post "" by method "GET" give response ""
Mar 27 15:26:10 symfony [info] {OAuth} call url "https://graph.facebook.com/me" with params "" post "" by method "GET" give response ""
Mar 27 15:26:10 symfony

here is my app.yml inputs

  melody:
    create_user: true               # to create a user for all melodies 
    facebook:
      key:       [KEY]
      secret:     [SECRET]
      scope:     [email]  
      user:
        username:                   # the name of a field of sfGuardUser
          call: me                  # api call
          path: id                  # path could be user.object.email for example to retrieve the right inforamtion
          prefix: Facebook_
        first_name:
          call: me
          path: first_name
        last_name: 
          call: me
          path: last_name
        email_address:  
          call: me  
          path: email  
          key: true                 # it's a key to retrieve user from other services based on this information
                                    # if no field is a key all are keys by default.  
    twitter:
      provider: twitter
      key:       [KEY]
      secret:    [SECRET]
      user:
        username:  
          call: me                  # api call
          path: id                  # path could be user.object.email for example to retrieve the right inforamtion
          prefix: Twitter_
        first_name:
          call: me
          path: screen_name
        email_address:  
          call: me  
          path: screen_name
          prefix: twitter_at_
          key: true                 # it's a key to retrieve user from other services based on this information
                                    # if no field is a key all are keys by default.

**Comment by mistrymj** [comment]
All,

I figured out the issue, doesn't have to do with the sfMelodyPlugin directly.

It is an issue with sfOAuth.class.php in sfDoctrineOAuthPlugin, in the call method.  This was caused by the cURL execution failing because the SSL cerificate could not be authenticated by a CA.  See here for a discussion: http://unitstep.net/blog/2009/05/05/using-curl-in-php-to-access-https-ssltls-protected-sites/

Hope this helps anyone having a similar issue.
